### PR TITLE
chore(linter): Migrate to eslint with typescript-eslint

### PR DESCRIPTION
**Pull request by JoseLion**
This PR migrates the linter from TSLint to use ESLint with `typescript-eslint`. A few improvements were added but also a few changes were required.

We didn't use ESLint before because 3 main reasons:
- Some of the most important TSLint rules were still missing.
- ESLint requires a lot more effort, configurations, plugins, and dev dependencies to work with TypeScript.
- There were some considerable performance issues with `typescript-eslint` compared to TSLint.

There are some points in favor of ESLint of course:
- In the end, it was only a matter of time until TSLint becomes too out-of-date regarding TypeScript versions, so we had to migrate at some point.
- ESLint and `typescript-eslint` gives us a lot more features: Use the type information to help with TS static checks, plugins, JS rules, etc.

Finally, the advantages of ESLint outweighed the disadvantages, so the extra effort is also justifiable.

**NOTE:** This PR resolves #73 by superceeding it.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -23,6 +23,7 @@ jobs:
           node-version: ${{ matrix.node }}
           cache: yarn
       - run: yarn install --immutable
+      - run: ./scripts/install-dev-on-example.sh
```
Comment (JoseLion) [inline .github/workflows/ci.yml:26]: Now we need to install the library and all the dependencies on the Example to accurately use the linter with type information

File: example/src/App.styles.ts [hunk 2]
```diff
@@ -1,4 +1,3 @@
-import { Colors } from "react-native/Libraries/NewAppScreen";
```
Comment (JoseLion) [inline example/src/App.styles.ts:1]: The type of `Colors` was resolved as `any`, so now `@typescript-eslint/no-explicit-any` rule complains

File: example/src/App.tsx [hunk 3]
```diff
@@ -37,30 +37,11 @@ export const App: React.FC = () => {
...
         </ButtonsGroupView>
       </SpotDescriptionView>
-    )
+    ),
   }, {
     alignTo: Align.SCREEN,
     position: Position.BOTTOM,
-    render: () => {
-      // You can also use the hook instead of the props here!
-      const { previous, next } = useSpotlightTour();
```
Comment (JoseLion) [inline example/src/App.tsx:46]: Moved the render function to its own component so the `react-hooks/rules-of-hooks` rule won't complain. The rule doesn't know the function is indeed a function component. We'll solve this later by deprecating the `render` property and changing its name to `Component`. Function which names start with an uppercase are considered components (by the rule, and by a React convention)

File: test/helpers/native-mocks.ts [hunk 4]
```diff
@@ -1,51 +1,38 @@
...
 import {
   Animated,
   MeasureInWindowOnSuccessCallback,
-  NativeMethods
+  NativeMethods,
 } from "react-native";
 
 import { MeasureOnSuccessCallbackParams } from "./measures";
 
-export function mockNativeComponent(modulePath: string, mockMethods: NativeMethods) {
```
Comment (JoseLion) [inline test/helpers/native-mocks.ts:10]: This helper function had to be refactored as it relied too much on `any` types, which caused the linter to heavily complain

File: example/src/App.tsx [hunk 5]
```diff
@@ -78,21 +59,18 @@ export const App: React.FC = () => {
...
           speed: 1,
           toValue: Dimensions.get("screen").height * 0.25,
-          useNativeDriver: false // Disabled as RNST does not use native driver either (for now)
+          useNativeDriver: false, // Translate animation not supported native by native driver
         })
-        .start(({ finished }) => finished
-          ? resolve()
-          : reject()
-        );
+        .start(() => resolve());
```
Comment (sebas1208) [inline example/src/App.tsx:73]: Is this change ok, not rejecting in the `finished = false` case?

**Review by sebas1208 — COMMENTED** [review]
LGTM!

File: example/src/App.tsx (hunk 5 above)
Comment (JoseLion) [reply example/src/App.tsx:73]: Yes, it's ok. The `finish = false` means that the animation could have been cancelled or restarted, it doesn't mean the animation went wrong. Animations are frequently cancelled by user interaction so we don't want to unnecessarily reject the promise 🙂

**Review by sebas1208 — COMMENTED** [review]
LGTM!

**Review by sebas1208 — APPROVED** [review]
LGTM!

**Pull request merged**